src/context/index.ts

```diff
@@ -0,0 +1,187 @@
+import { loadLogs } from './loadLogs'
+import { Panel } from './panel'
+import { regionToSelection } from './regionToSelection'
+import { ResultDiagnostic } from './resultDiagnostic'
+import { Store } from './store'
+
+export async function activate(context: ExtensionContext) {
+	const disposables = context.subscriptions
+	Store.extensionPath = context.extensionPath
+	Store.globalState = context.globalState
+	disposables.push(commands.registerCommand('sarif.clearState', () => {
+		context.globalState.update('view', undefined)
+		commands.executeCommand('workbench.action.reloadWindow')
+	}))
+	const store = new Store()
+
+	// Basing
+	const urisNonSarif = await workspace.findFiles('**/*', '.sarif') // Ignore folders?
+	const fileAndUris = urisNonSarif.map(uri => [uri.path.split('/').pop(), uri.path]) as [string, string][]
+	const baser = new Baser(mapDistinct(fileAndUris), store)
+
+	// Panel
+	const panel = new Panel(context, baser, store)
+	disposables.push(commands.registerCommand('sarif.showPanel', () => panel.show()))
+
+	// Suggest In-Project Sarif Files
+	;(async () => {
+		const urisSarifInWorkspace = await workspace.findFiles('**/*.sarif', '.sarif/**/*.sarif')
+		const count = urisSarifInWorkspace.length
+		if (!count) return
+		if (await window.showInformationMessage(`Discovered ${count} SARIF logs in your workspace.`, 'View in SARIF Panel')) {
+			store.logs.push(...await loadLogs(urisSarifInWorkspace))
+			panel.show()
+		}
+	}) // Disabled while we evaluate the future of this feature.
+
+	// Diagnostics
+	const diagsAll = languages.createDiagnosticCollection('SARIF')
+	const setDiags = (doc: TextDocument) => {
+		if (doc.fileName.endsWith('.git')) return
+		const artifactPath = baser.translateLocalToArtifact(doc.uri.path)
+		const severities = {
+			error: DiagnosticSeverity.Error,
+			warning: DiagnosticSeverity.Warning,
+		} as Record<string, DiagnosticSeverity>
+		const diags = store.results
+			.filter(result => result._uri === artifactPath)
+			.map(result => new ResultDiagnostic(
+				regionToSelection(doc, result._region),
+				result._message ?? '—',
+				severities[result.level ?? ''] ?? DiagnosticSeverity.Information, // note, none, undefined.
+				result,
+			))
+		diagsAll.set(doc.uri, diags)
+	}
+	workspace.textDocuments.forEach(setDiags)
+	workspace.onDidOpenTextDocument(setDiags)
+	workspace.onDidCloseTextDocument(doc => diagsAll.delete(doc.uri)) // Spurious *.git deletes don't hurt.
+	observe(store.logs, () => workspace.textDocuments.forEach(setDiags))
+
+	// Open Documents <-sync-> Store.logs
+	const syncActiveLog = async (doc: TextDocument) => {
+		if (!doc.fileName.match(/\.sarif$/i)) return
+		if (store.logs.some(log => log._uri === doc.uri.toString())) return
+		store.logs.push(...await loadLogs([doc.uri]))
+		panel.show()
+	}
+	workspace.textDocuments.forEach(syncActiveLog)
+	workspace.onDidOpenTextDocument(syncActiveLog)
+	workspace.onDidCloseTextDocument(doc => {
+		if (!doc.fileName.match(/\.sarif$/i)) return
```

Comment:
>doc [](start = 7, length = 3)

You can use the doc language id to look for the language defined in package.json

Reply:
The doc.languageId is currently just "json". I guess we can defined a new language id 'sarif'. But given we only have 1 file extension (and not all JSON files are SARIF files), and we don't have any special language features (above and beyond JSON), is there any other benefit to going this direction?

---
In reply to: [435424162](https://github.com/microsoft/sarif-vscode-extension/pull/279#discussion_r435424162) [](ancestors = 435424162)